### PR TITLE
[WIP] qos: list rule types

### PR DESCRIPTION
**Pull request by Twister915**
For #1027, Implements listing rule types (`GET /rule-types`)

Links to the line numbers/files in the OpenStack source code that support the
code in this PR:

https://github.com/openstack/neutron/blob/master/neutron/objects/qos/rule_type.py#L61

WIP, implementing other routes

### CODE REVIEW

**Comment by jtopjian** [comment]
@Twister915 Thanks for working on this. Let me know when this PR is ready for a review :)

> WIP, implementing other routes

It might be beneficial to do a few PRs individually in order to get familiar with certain coding styles. This might help reduce multiple changes across PRs.

For example, this PR is really good. There are a few areas that need touched up, though. I'm more than happy to point them out but only if it's ready :)

**Comment by Twister915** [comment]
@jtopjian thanks! happy to help 👍  

I was planning on implementing each route in a different PR, because that's what the contributor guide said to do. Is that what you mean by doing them individually? Would you rather me put them in one PR after I get the hang of it?

Also... I'm pretty sure I'm done with this specific route, so feel free to review it before I start submitting further PRs for other routes. I have [WIP] in the title because I assume I wouldn't want to merge this route without merging the other ones.

**Comment by jtopjian** [comment]
> I was planning on implementing each route in a different PR, because that's what the contributor guide said to do

Yes, this is correct.

> Would you rather me put them in one PR after I get the hang of it?

Still one PR per route/call, please :)

> I have [WIP] in the title because I assume I wouldn't want to merge this route without merging the other ones.

This. It's totally fine to merge individual routes/calls without having the entire set ready, especially in the beginning and for new contributors. This helps both the contributor and reviewer get a better feel of the resource. There's usually a good amount of changes done in the beginning but things settle down after a few PRs and then you can easily open multiple PRs without concern of doing a large chain of rebasing.

File: openstack/networking/v2/extensions/qos/results.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+	// now make the out slice
+	out = make([]string, len(list.RuleTypes))
+	// and copy the result to the return value
+	for i, t := range list.RuleTypes {
+		out[i] = t.Type
+	}
+
+	return
+}
```
Comment (jtopjian) [inline openstack/networking/v2/extensions/qos/results.go]: I've debated this one for almost an hour. I understand that the results of retrieving the details of a rule type are very different than the list request and that the list request is only returning a `Type` field, but I think we should create a proper struct for `RuleType`, even though it'll only contain one field.

The more detailed result can be called `RuleTypeDetails`. Or choose names you feel are suitable.

I realize that this basically rewrites this entire file and I apologize about that. But I have a feeling this is better in the long-run. To help get you started, the contents of the file would look something like:

```go
type RuleType struct {
  Type string `json:"type"`
}

// The result of listing the qos rule types
type ListRuleTypesPage struct {
  pagination.SinglePageBase
}

// IsEmpty indicates whether a RuleTypesPage is empty.
func (r ListRuleTypesPage) IsEmpty() (bool, error) {
  v, err := ExtractRuleTypes(r)
  return len(v) == 0, err
}

func ExtractRuleTypes(r pagination.Page) ([]RuleType, error) {
  var s struct {
    RuleTypes []RuleType `json:"rule_types"`
  }
  err := (r.(ListRuleTypesPage)).ExtractInto(&s)
  return s.RuleTypes, err
}
```

File: openstack/networking/v2/extensions/qos/requests.go [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package qos
+
+import "github.com/gophercloud/gophercloud"
+
+// ListRuleTypes returns the list of rule types from the server
+func ListRuleTypes(c *gophercloud.ServiceClient) (result ListRuleTypesResult) {
+	_, result.Err = c.Get(listRuleTypesURL(c), &result.Body, nil)
+	return
+}
```
Comment (jtopjian) [inline openstack/networking/v2/extensions/qos/requests.go]: With the changes made to `results.go`, this function will now look something like:

```go
// ListRuleTypes retrieves the QoS rule types.
func ListRuleTypes(client *gophercloud.ServiceClient) pagination.Pager {
  url := listRuleTypesURL(c)

  return pagination.NewPager(client, url, func(r pagination.PageResult) pagination.Page {
    return ListRuleTypesPage{pagination.SinglePageBase(r)}
  })
}
```

If you were to put this in a `qos/ruletypes` package, the function name could be changed to just `List`.

File: openstack/networking/v2/extensions/qos/testing/fixtures.go [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+        },
+        {
+            "type": "dscp_marking"
+        },
+        {
+            "type": "minimum_bandwidth"
+        }
+    ]
+}`
+)
```
Comment (jtopjian) [inline openstack/networking/v2/extensions/qos/testing/fixtures.go]: nit:

```go
const ListRuleTypesResponse = `
{
    "rule_types": [
    ...
}
`
```

**Comment by Twister915** [comment]
When submitting my next pull request, should I base it off this branch? Also, should we merge this to master before merging the next pull request? The diff between master and the pull request I want to make includes all the changes from this pull request, making it difficult to review (presumably).

**Comment by Twister915** [comment]
@jtopjian When submitting my next pull request, should I base it off this branch? Also, should we merge this to master before merging the next pull request? The diff between master and the pull request I want to make includes all the changes from this pull request, making it difficult to review (presumably).

**Comment by jtopjian** [comment]
> When submitting my next pull request, should I base it off this branch? 

Totally your preference. I prefer the "chaining method" where the next PR's branch will be based off of this branch in anticipation of this PR being merged.

Some people have based all PRs off of a common master. This causes duplication between PRs, but it's a style they prefer.

> Also, should we merge this to master before merging the next pull request? 

Yes, I'll have some time for another review later today or tomorrow. If all looks good, I'll merge it.

When chaining PRs, you have to factor in a parent PR needing changes. Then you have to rebase all the way down the chain. This is why I recommended not opening too many PRs at first :) once you get a feel for the style and such it's easier to have multiple PRs open without worry of changes required. 

If your PRs do not require any changes and you're chaining branches from one to another, Github will automatically recognize this and cleanly merge one after another even though child PRs show multiple commits. It's a little confusing -- let me know if you have any questions :)

**Pull request merged**